File: java/getHint.java
class Solution {
    public String getHint(String secret, String guess) {
        HashMap<Character, Integer> freq = new HashMap();
        int A = 0;
        int B = 0;

        for (int i = 0; i < secret.length(); i++) {
            Character currentChar = secret.charAt(i);

            if (currentChar == guess.charAt(i)) {
                A++;
            } else {
                if (freq.containsKey(currentChar)) {
                    int count = freq.get(currentChar);
                    freq.put(currentChar, count + 1);
                } else {
                    freq.put(currentChar, 1);
                }
            }
        }

        for (int i = 0; i < guess.length(); i++) {
            if (guess.charAt(i) != secret.charAt(i)) {
                Character currentChar = guess.charAt(i);

                // check if char exist
                boolean charExist = freq.containsKey(currentChar);
                
                if (charExist) {
                    B++;
                    int count = freq.get(currentChar);

                    if (count == 1) {
                        freq.remove(currentChar);
                    } else {
                        freq.put(currentChar, count - 1);
                    }
                }
            }
        }

        return A + "A" + B + "B";
    }
}

File: js/getHint.js
/**
 * @param {string} secret
 * @param {string} guess
 * @return {string}
 */
var getHint = function(secret, guess) {
    const freq = {};
    let A = 0
    let B = 0;

    // check for A(s)
    for (let i = 0; i <guess.length; i++) {
        if (guess[i] === secret[i]) {
            A++;
        } else {
            freq[secret[i]] = (freq[secret[i]] ?? 0) + 1;
        }
    }

    // check for B(s)
    for (let i = 0; i < guess.length; i++) {
        if (guess[i] !== secret[i]) {
            const currentChar = guess[i]
            const charExists = freq[currentChar] !== undefined;

            if (charExists) {
                B++;
            }

            const currentFreq = freq[currentChar];
            if (currentFreq !== undefined) {
                if (currentFreq === 1) {
                    delete freq[currentChar]
                } else {
                    freq[currentChar] = freq[currentChar] - 1
                }

            }
        }
    }

    return `${A}A${B}B`;
};